lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+			evt.message = node->message;
+			notify_event(&evt);
+			memset(&node->message, 0, sizeof(struct qos_data));
+			sys_slist_remove(&ctx.pending_list, &node_prev->header, &node->header);
+			return 0;
+		}
+
+		node_prev = node;
+	};
+
+	return -ENODATA;
+}
+
+/* Public API functions */
+
+int qos_init(qos_evt_handler_t evt_handler)
+{
+	if (evt_handler == NULL) {
+		LOG_DBG("Handler is NULL");
+		return -EINVAL;
+	}
+
+	if (ctx.initialized) {
+		return -EALREADY;
+	}
+
+	ctx.initialized = true;
+
+	LOG_DBG("Registering handler %p", evt_handler);
+	ctx.app_evt_handler = evt_handler;
+
+	/* Initializing linked list and delayed work. */
+	sys_slist_init(&ctx.pending_list);
+	k_work_init_delayable(&ctx.timeout_handler_work, timeout_handler_work_fn);
+	return 0;
+}
+
+int qos_message_add(struct qos_data *message)
+{
+	int err = 0, ret;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_NEW,
+		.message = *message
+	};
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	if (message == NULL) {
+		LOG_WRN("Message cannot be NULL");
+		err = -EINVAL;
+		goto exit;
+	}
+
+	/* Only ACK_REQUIRED messages are added to the internal list. */
+	if (qos_message_has_flag(message, QOS_FLAG_RELIABILITY_ACK_REQUIRED)) {
+		ret = list_append(message);
+		if (ret < 0) {
+			LOG_WRN("No list entries available, error: %d", ret);
+			evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+			notify_event(&evt);
+			err = -ENOMEM;
+			goto exit;
+		}
+
+		ctx.list_internal[ret].message.notified_count++;
+		notify_event(&evt);
+	} else {
+		notify_event(&evt);
+
+		/* If the message does not carry the ACK_REQUIRED flag, its removed from the
+		 * internal list immediately.
+		 */
+		evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+		notify_event(&evt);
+	}
+
+	timer_start();
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+int qos_message_remove(uint32_t id)
+{
+	int err;
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	err = list_remove(id);
+	if (err) {
+		goto exit;
+	}
+
+	/* If the removed message is the last in the pending list, we stop the internal timer
+	 * and clear the retry counter.
+	 */
+	if (sys_slist_is_empty(&ctx.pending_list)) {
+		LOG_WRN("QoS list is empty!");
+		ctx.retry_count = 0;
+		k_work_cancel_delayable(&ctx.timeout_handler_work);
+	}
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+void qos_message_print(const struct qos_data *message)
+{
+	LOG_DBG("Notified count: %d", message->notified_count);
+	LOG_DBG("Message heap_allocated: %d", message->heap_allocated);
+	LOG_DBG("Message ID: %d", message->id);
+	LOG_DBG("Message Buffer pointer: %p", message->data.buf);
+	LOG_DBG("Message Buffer length: %d", message->data.len);
+	LOG_DBG("Message Flags: %x", message->flags);
+	LOG_DBG("Message type: %d", message->type);
+}
+
+bool qos_message_has_flag(const struct qos_data *message, uint32_t flag)
+{
+	return ((message->flags & flag) == flag) ? true : false;
+}
+
+uint16_t qos_message_id_get_next(void)
+{
+	if (message_id_next > UINT16_MAX) {
```

Comment:
I don't think it'll ever be true, as `message_id_next` is `uint16_t`.
btw, any particular reason to use the base value? Can't we just randomize the initial value and allow to overflow the ID?

Reply:
😧 You're right. Ill change it to `==` instead. I originally used `uint32_t` so its leftover from that.

I had to use a base value because nRF Cloud mqtt library reserves certain msg ID values for internal use. I noticed that if specific value were used the corresponding ACK would never get propagated to the application.

https://github.com/simensrostad/fw-nrfconnect-nrf/blob/40a282ea9973e5e55590f138f7a2a794cd8c734b/include/net/nrf_cloud.h#L21